library(argparse)
library(data.table)

# tag_file='fastq/cluster_analy/chipseq_xing_ublocks_ublocks_filt10.tsv',
# wkd <- 'fastq/kundaje_encode/cluster_analy'
# rd_fn <- file.path(wkd,sprintf('%s_heatmap.rd',exp_label))

if (F) {
  args_tmp <- commandArgs(trailingOnly = F)
  scriptPath <- dirname(normalizePath(sub("--file=","",args_tmp[grep("--file",args_tmp)])))
  
  parser <- ArgumentParser(description='prepropa')
  
  parser$add_argument("-i", "--input_tsv", type="character", required=TRUE,
                      dest="input_tsv",
                      help="input TSV file derived from a05")
  
  parser$add_argument("-c", "--cluster_rd", type="character", required=TRUE,
                      dest="cluster_rd",
                      help="cluster analysis rd file generated by b04_gen_heatmap_filtw.r")
  
  parser$add_argument("-r", "--reuse", type="integer", required=FALSE,
                      dest="reuse",default=1,
                      help="Yes:[1], No:0")
  args <- parser$parse_args()
} else {
  args <- data.table(chipseq_manorm_rd='fastq/kundaje_encode/goi_apa_manorm.rd',
                     cluster_rd='fastq/cluster_analy/chipseq_cluster.rd',
                     wkd='fastq/cluster_analy',
                     r=0)
}

library(openxlsx)
library(handy)
source(file.path(Sys.getenv('R_UTIL'),'lib_util.R'))
source(file.path(Sys.getenv('R_UTIL'),'lib_apps.R'))
source(file.path(Sys.getenv('R_UTIL'),'lib_chipseq.r'))

# ------------------------------
apa_call <- 'ShorterInB'
# ----------------------------
message('loading st3 ...')

apa_colon_xlsx <- "../01_polyAseq/01_wkd/out/04_AnnoApa/output/apa.ann_HCT116_vs_DKO.xlsx"
stopifnot(file.exists(apa_colon_xlsx))
apa <- read_supp_tab3(apa_colon_xlsx,sheet_i=2) #keep apa to bring annotation back
apa.gr <- makeGRanges(apa[call==apa_call,])

# ----------------------------
message('loading clustering and mean depth of MBD ...')

# md_dt_cls <- fread(args$input_tsv)
load(args$cluster_rd) #md_dt_cls MBD is already in log2

# ----------------------------
message('a simple analysis of 3 genes of interest')

library(reshape2)
goi3 <- md_dt_cls[gene %in% c('NFYA','DNAJB6','HEATR2') & width>0,]
goi3.gcw <- goi3[,.(total_bp=sum(width)),by=c('gene','cluster_id')]
goi3.w <- dcast(goi3.gcw, cluster_id ~ gene, value.var="total_bp")
goi3.w[is.na(goi3.w)]<-0.
rownames(goi3.w) <- paste0('cluster_',goi3.w$cluster_id)
goi3.w$cluster_id <- NULL
apply(goi3.w,2,norm<-function(x){return (x/sum(x))})

#             DNAJB6     HEATR2      NFYA
# cluster_1 0.006952112 0.05551481 0.0000000
# cluster_2 0.001063264 0.26738046 0.0000000
# cluster_3 0.402977140 0.17864476 0.0000000
# cluster_4 0.103586472 0.07890877 0.5325444
# cluster_5 0.000000000 0.02383397 0.0000000
# cluster_6 0.063223326 0.01166031 0.0000000
# cluster_7 0.005602585 0.03512760 0.3060785
# cluster_8 0.038604670 0.14623057 0.0000000
# cluster_9 0.377990431 0.20269874 0.1613771

mbd_dt <- md_dt_cls[,list(chr,start,end,HCT116_Mbd,DKO_Mbd)]
mbd_dt$HCT116_Mbd <- 2^(mbd_dt$HCT116_Mbd) - 1
mbd_dt$DKO_Mbd <- 2^(mbd_dt$DKO_Mbd) - 1

md_dt_cls <- md_dt_cls[,list(chr,start,end,cluster_id,pos)]
colnames(md_dt_cls) <- c('chr','start','end','cluster_id','pos')
md_dt_cls <- md_dt_cls[order(chr,start,end),]
clustered.gr <- makeGRanges(md_dt_cls)
rm(md_dt_cls)

setnames(mbd_dt,
         c('HCT116_Mbd','DKO_Mbd'),
         c('normalized_read_density_in_HCT116_MBD','normalized_read_density_in_DKO_MBD'))

# --------------------------
message('annotate gene of interest to cluster table...')
ij <- findOverlaps(clustered.gr,apa.gr)
clustered.gr$gene[queryHits(ij)] <- apa.gr$gene_name[subjectHits(ij)]

# ----------------------
message('to retrieve normalized mean depth from MANorm report')

load(args$chipseq_manorm_rd) #manorms
manorms <- lapply(manorms,function(manorm){return(manorm$dt)})
manorms[['HCT116_DKO_MBD']] <- mbd_dt

# -------------
message('storing signals into a matrix format splitted by two groups (control vs. expr) ...')
M <- length(clustered.gr)
N <- length(manorms)

jmessage('prep a matrix to hold binding depth ...')
T2 <- matrix(0., nrow = M, ncol = 2*N)
labels <- as.data.table(tstrsplit(names(manorms),'_'))
labels$names <- names(manorms)
colnames(labels) <- c('ctrl','expr','protein','name')

chipseq_tags <- rep(NA,2*N)

for (i in 1:N) {
  ctrl_tag <- sprintf('%s_%s',labels[i,ctrl],labels[i,protein])
  ctrl_col <- sprintf('normalized_read_density_in_%s',ctrl_tag)
  
  expr_tag <- sprintf('%s_%s',labels[i,expr],labels[i,protein])
  expr_col <- sprintf('normalized_read_density_in_%s',expr_tag)
  
  md.gr <- makeGRanges(manorms[[i]][,list(chr,start,end,ctrl=get(ctrl_col),expr=get(expr_col))])
  
  ij <- findOverlaps(clustered.gr,md.gr)
  T2[queryHits(ij),i] <- md.gr$ctrl[subjectHits(ij)]
  chipseq_tags[[i]] <- ctrl_tag
  
  T2[queryHits(ij),(N+i)] <- md.gr$expr[subjectHits(ij)]
  chipseq_tags[[N+i]] <- expr_tag
}

md_dt <- as.data.table(log2(T2+1.))
colnames(md_dt) <- chipseq_tags
md_dt$cluster_id <- clustered.gr$cluster_id
md_dt$gene <- clustered.gr$gene
md_dt$width <- width(clustered.gr)
md_dt$pos <- clustered.gr$pos

library(ggplot2)
library(gplots)
library(factoextra)
library(RColorBrewer)
library(gridExtra)

# ================

sum_tab <- md_dt[,.(num_of_genomic_region=.N),by='cluster_id'][order(cluster_id)]
tbp <- md_dt[,.(genomic_region_total_bp=sum(width)),by='cluster_id'][order(cluster_id)]
genes <- md_dt[,.(num_of_genes=length(unique(gene))),by='cluster_id'][order(cluster_id)]

wb <- createWorkbook()

sum_tab <- merge(sum_tab,tbp,by='cluster_id')
sum_tab <- merge(sum_tab,genes,by='cluster_id')

addWorksheet(wb,sheetName = 'basic stats of clusters')
writeData(wb,sheet='basic stats of clusters',x=sum_tab)

addWorksheet(wb,sheetName = 'genes per cluster')
gene_summary_tab <- md_dt[,.(count=.N,total_bp=sum(width)),by=c('cluster_id','gene')][order(cluster_id,gene)]
writeData(wb,sheet='genes per cluster',x=gene_summary_tab)

pdf_file <- file.path(args$wkd,'fig03d_cluster_analysis.pdf')

md_dt_cls <- split(md_dt,by='cluster_id')

md_dt_cls <- md_dt_cls[order(names(md_dt_cls))]

cohesin <- c('CTCF','RAD21','SMC1')
histone <- c('H3K27Ac','Pol2Ser2','Pol2Ser5')
methyl <- c('MBD')

#chipseq_tags <- c('CTCF','RAD21','SMC1','H3K27Ac','Pol2Ser2','Pol2Ser5')
chipseq_tags <- c(cohesin,histone,methyl)
cluster_names <- names(md_dt_cls)
# pdf(file=pdf_file)

I <- length(md_dt_cls)
J <- length(chipseq_tags)
p1s <- vector("list",I*J)
kl_stats <- matrix(0,nrow=J,ncol=I)
chi2_stats <- matrix(0,nrow=J,ncol=I)
wx_stats <- matrix(0,nrow=J,ncol=I)

for (i in 1:I) { #for each cluster
  md_dt_clsj <- md_dt_cls[[i]][,c(1:14,18)]
  md_dt_clsjl <- melt(md_dt_clsj,id.vars ="pos")
  setnames(md_dt_clsjl,c('variable','value'),c('group_protein','log2x'))
  
  group_proteins <- tstrsplit(md_dt_clsjl$group_protein,'_')
  md_dt_clsjl$group <- group_proteins[[1]]
  md_dt_clsjl$protein <- group_proteins[[2]]
  
  cli <- as.integer(cluster_names[[i]])
  dim(md_dt[cluster_id == cli,])[1]
  
  for (j in 1:J) {
    protein_j <- chipseq_tags[[j]]
    dtj <- subset(md_dt_clsjl,protein==protein_j)
    p1 <- ggplot(dtj,aes(x=log2x,linetype=group,color=group))
    p1 <- p1 + geom_line(stat='density')
    #p1 <- p1 + scale_fill_manual(values=c(,'#ffbf00','#0000ff'))
    p1 <- p1 + scale_linetype_manual(values=c("solid","dashed"))
    p1 <- p1 + scale_color_manual(values=c('#ffbf00','#0000ff'))
    p1 <- p1 + ylim(0,1)
    p1 <- p1 + xlim(0,10)
    p1 <- p1 + labs(x='log2(depth)',y='density')
    # p1 <- p1 + labs(x='depth',
    #               y='density',
    #               title=sprintf('cid[%s],%s',cluster_names[i],protein_j))
    p1 <- p1 + theme(legend.position="none")
    p1 <- p1 + handy::ggnice()
    cls_protein <- sprintf('%d_%s',i,protein_j)
    ij <- i + I*(j-1)
    message(sprintf("i:%d,j:%d,cls_protein:%s,ij:%d",i,j,cls_protein,ij))
    p1s[[ij]] <- p1
    
    p <- ggplot(dtj)+geom_density(aes(x=log2x,color=group))
    pd <- as.data.table(ggplot_build(p)$data[[1]])
    pd_by_group <- split(pd,by='group')
    X <- pd_by_group[[1]]$density
    Y <- pd_by_group[[2]]$density
    
    if (F) {
      kl_stats[j,i] <- KL.plugin(X,Y)
      chi2_stats[j,i] <- 0.5*chi2.plugin(X,Y)
    } else {
      dtj_by_group <- split(dtj,by='group')
      dtj_by_group[[1]]$log2x
      dtj_by_group[[2]]$log2x
      
      tag_comp <- data.table(HCT116=dtj_by_group$HCT116$log2x,
                             DKO=dtj_by_group$DKO$log2x)
      
      wtest <- wilcox.test(tag_comp$HCT116, tag_comp$DKO, data=tag_comp)
      wx_stats[j,i] <- wtest$p.value
    }
  }
}
grid.arrange(grobs=p1s, nrow=J, ncol = I)
x <- 1

if (F) {
  addWorksheet(wb,sheetName = 'HCT116_vs_DKO_KL_divergence')
  kl_stats<-as.data.table(kl_stats)
  colnames(kl_stats) <- cluster_names
  kl_stats$protein <- chipseq_tags
  writeData(wb,sheet='HCT116_vs_DKO_KL_divergence',x=kl_stats)
  
  addWorksheet(wb,sheetName = 'HCT116_vs_DKO_chi2')
  chi2_stats <- as.data.table(chi2_stats)
  colnames(chi2_stats) <- cluster_names
  chi2_stats$protein <- chipseq_tags
  writeData(wb,sheet='HCT116_vs_DKO_chi2',x=chi2_stats)
}

addWorksheet(wb,sheetName = 'HCT116_vs_DKO_wilcox')
wx_stats <- as.data.table(wx_stats)
colnames(wx_stats) <- cluster_names
wx_stats$protein <- chipseq_tags
writeData(wb,sheet='HCT116_vs_DKO_wilcox',x=wx_stats)

xlsx_file <- file.path(args$wkd,'fig03d_cluster_analy.xlsx')
saveWorkbook(wb,xlsx_file,overwrite=TRUE)

# ---------------------------------------
message('protein wise')
groups <- c('HCT116','DKO')
md_dt2 <- md_dt[,c(1:14,18)]

md_dt_melt <- melt(md_dt2,id.vars ="pos")
setnames(md_dt_melt,c('variable','value'),c('group_protein','log2x'))

group_proteins <- tstrsplit(md_dt_melt$group_protein,'_')
md_dt_melt$group <- group_proteins[[1]]
md_dt_melt$protein <- group_proteins[[2]]
md_dt_melt$cluster_id <- as.factor(md_dt[match(md_dt_melt$pos,md_dt$pos),cluster_id])

md_dt_by_group <- split(md_dt_melt,by='group')
sample_groups <- names(md_dt_by_group)

k9_pal <- c('#191970','#006400','#ff4500','#ffd700','#00ff00','#00bfff','#0000ff','#ff69b4','#ffe4c4')

for (i in 1:length(md_dt_by_group)) {
  md_dt_by_proteins <- split(md_dt_by_group[[i]],by='protein')
  p1s <- list()
  proteins <- names(md_dt_by_proteins)
  for (j in 1:length(md_dt_by_proteins)) {
    
    p1 <- ggplot(md_dt_by_proteins[[j]])
    p1 <- p1 + geom_density(aes(x=log2x,color=cluster_id))
    p1 <- p1 + scale_color_manual(values=k9_pal)
    p1 <- p1 + ylim(0,1)
    p1 <- p1 + xlim(0,10)
    p1 <- p1 + labs(x='depth',
                  y='density',
                  title=sprintf('%s,%s',sample_groups[[i]],proteins[[j]]))
    p1 <- p1 + theme(legend.position="right")
    p1 <- p1 + handy::ggnice()
    cls_protein <- sprintf('%d_%s',i,proteins[[j]])
    p1s[[cls_protein]] <- p1
  }
  grid.arrange(grobs=p1s, nrow=2)
}

dev.off()